File: algorithms/2_medium/N/01976_number_of_ways_to_arrive_at_destination/heap.cpp
constexpr int64_t MOD = 1'000'000'007;

class Solution {
    struct Cmp {
        bool operator()(const pair<int64_t, int64_t>& a, const pair<int64_t, int64_t>& b) {
            return a.first > b.first;
        }
    };
public:
    static int countPaths(const int n, const vector<vector<int>>& roads) {
        assert(n > 0 && n <= 200);
        auto adj = vector<vector<pair<int, int>>>(n, vector<pair<int, int>>());
        for (const auto& r : roads) {
            assert(r.size() == 3);
            const int u = r[0];
            const int v = r[1];
            const int w = r[2];
            assert(u >= 0 && u < n && v >= 0 && v < n && w >= 1);
            adj[u].push_back({v, w});
            adj[v].push_back({u, w});
        }

        auto heap = priority_queue<
            pair<int64_t, int64_t>,
            vector<pair<int64_t, int64_t>>,
            Cmp
        >();
        auto costs = vector<pair<int64_t, int64_t>>(n, { numeric_limits<int64_t>::max(), 0 });
        heap.push({0, 0});
        costs[0] = {0, 1};

        while (!heap.empty()) {
            const auto [cost, u] = heap.top();
            heap.pop();
            if (cost > costs[u].first) { continue; }

            const int64_t ways = costs[u].second;

            for (const auto [v, w] : adj[u]) {
                const int64_t new_cost = cost + w;
                if (new_cost < costs[v].first) {
                    costs[v].first = new_cost;
                    costs[v].second = ways;
                    heap.push({new_cost, v});
                } else if (new_cost == costs[v].first) {
                    costs[v].second = (costs[v].second + ways) % MOD;
                }
            }
        }

        return static_cast<int>(costs[(n - 1)].second);
    }
};

File: algorithms/2_medium/N/01976_number_of_ways_to_arrive_at_destination/heap.rs
use std::collections::BinaryHeap;
use std::cmp::Reverse;

impl Solution {
    const MOD: i64 = 1_000_000_007;

    pub fn count_paths(n: i32, roads: Vec<Vec<i32>>) -> i32 {
        assert!(n > 0 && n <= 200);
        let mut adj: Vec<Vec<(i32, i32)>> = vec![Vec::default(); n as usize];
        for r in roads {
            assert!(r.len() == 3);
            let (u, v, w) = (r[0], r[1], r[2]);
            assert!(u >= 0 && u < n && v >= 0 && v < n && w >= 1);
            adj[u as usize].push((v, w));
            adj[v as usize].push((u, w));
        }

        let mut heap: BinaryHeap<Reverse<(i64, i32)>> = BinaryHeap::with_capacity(2 * n as usize);
        let mut costs = vec![(i64::MAX, 0_i64); n as usize];
        heap.push(Reverse((0, 0)));
        costs[0] = (0, 1);

        while let Some(Reverse((cost, u))) = heap.pop() {
            if cost > costs[u as usize].0 { continue; }

            let ways = costs[u as usize].1;

            for &(v, w) in &adj[u as usize] {
                let new_cost = cost + w as i64;
                if new_cost < costs[v as usize].0 {
                    costs[v as usize].0 = new_cost;
                    costs[v as usize].1 = ways;
                    heap.push(Reverse((new_cost, v)));
                } else if new_cost == costs[v as usize].0 {
                    costs[v as usize].1 = (costs[v as usize].1 + ways) % Self::MOD;
                }
            }
        }

        costs[(n - 1) as usize].1 as i32
    }
}
